src/test/java/redis/clients/jedis/tests/SSLJedisClusterTest.java

```diff
@@ -0,0 +1,262 @@
+      Assert.assertEquals("The JedisConnectionException does not contain the expected message.",
+          "Could not get a resource from the pool", e.getMessage());
+    }
+    jc.close();
+
+    JedisCluster jc2 = null;
+    try {
+      jc2 = new JedisCluster(new HostAndPort("127.0.0.1", 8379), DEFAULT_TIMEOUT, DEFAULT_TIMEOUT,
+          DEFAULT_REDIRECTIONS, "cluster", null, DEFAULT_CONFIG, true, 
+          null, null, hostnameVerifier, portMap);
+      Assert.fail("The code did not throw the expected NullPointerException.");
+    } catch (NullPointerException e) {
+      // Null pointer exception occurs from closing Jedis object that did not connect due to custom
+      // hostname validation. When closing this Jedis object, the RedisOutputStream in the underlying 
+      // Connection object is null and causes this exception
+    }
+    if (jc2 != null) {
+      jc2.close();
+      Assert.fail("JedisCluster object is unexpectedly non-null");
+    }
+    
+    JedisCluster jc3 = new JedisCluster(new HostAndPort("localhost", 8379), DEFAULT_TIMEOUT, DEFAULT_TIMEOUT,
+        DEFAULT_REDIRECTIONS, "cluster", null, DEFAULT_CONFIG, true, 
+        null, null, localhostVerifier, portMap);;
+    jc3.get("foo");
+    jc3.close();
+  }
+  
+  @Test
+  public void connectWithCustomSocketFactory() throws Exception {
+    final SSLSocketFactory sslSocketFactory = SSLJedisTest.createTrustStoreSslSocketFactory();
+
+    JedisCluster jc = new JedisCluster(new HostAndPort("localhost", 8379), DEFAULT_TIMEOUT, DEFAULT_TIMEOUT,
+                                       DEFAULT_REDIRECTIONS, "cluster", null, DEFAULT_CONFIG, true, 
+                                       sslSocketFactory, null, null, portMap);
+    assertEquals(3, jc.getClusterNodes().size());
+    jc.close();
+  }
+  
+  @Test
+  public void connectWithEmptyTrustStore() throws Exception {
+    final SSLSocketFactory sslSocketFactory = SSLJedisTest.createTrustNoOneSslSocketFactory();
+
+    JedisCluster jc = null;
+    try {
+      jc = new JedisCluster(new HostAndPort("localhost", 8379), DEFAULT_TIMEOUT, DEFAULT_TIMEOUT,
+            DEFAULT_REDIRECTIONS, "cluster", null, DEFAULT_CONFIG, true, 
+            sslSocketFactory, null, null, null);
+      Assert.fail("The code did not throw the expected JedisConnectionException.");
+    } catch (JedisConnectionException e) {
+      Assert.assertEquals("Unexpected first inner exception.",
+          SSLException.class, e.getCause().getClass());
+      Assert.assertEquals("Unexpected second inner exception.",
+          SSLException.class, e.getCause().getCause().getClass());
+      Assert.assertEquals("Unexpected third inner exception",
+          RuntimeException.class, e.getCause().getCause().getCause().getClass());
+      Assert.assertEquals("Unexpected fourth inner exception.",
+          InvalidAlgorithmParameterException.class, e.getCause().getCause().getCause().getCause().getClass());
+    }
+    
+    if (jc != null) {
```

Comment:
If `jc != null`, execution will not reach this point because of previous `Assert.fail()`. Put this block in finally and remove following `Assert.fail()`.